pkg/config/decoder/decoder.go

```diff
@@ -0,0 +1,150 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package decoder
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/talos-systems/talos/pkg/config"
+)
+
+var (
+	// ErrMissingVersion indicates that the manifest is missing a version.
+	ErrMissingVersion = errors.New("missing version")
+	// ErrMissingKind indicates that the manifest is missing a kind.
+	ErrMissingKind = errors.New("missing kind")
+	// ErrMissingSpec indicates that the manifest is missing a spec.
+	ErrMissingSpec = errors.New("missing spec")
+)
+
+const (
+	// ManifestVersionKey is the string indicating a manifest's version.
+	ManifestVersionKey = "version"
+	// ManifestKindKey is the string indicating a manifest's kind.
+	ManifestKindKey = "kind"
+	// ManifestSpecKey is represents a manifest's spec.
+	ManifestSpecKey = "spec"
+	// ManifestDeprecatedKey is represents the deprected v1alpha1 manifest.
+	ManifestDeprecatedKey = "machine"
+)
+
+// Decoder represents a multi-doc YAML decoder.
+type Decoder struct {
+	source []byte
+}
+
+// Decode decodes all known manifests.
+func (d *Decoder) Decode() ([]interface{}, error) {
+	return d.decode()
+}
+
+// NewDecoder initializes and returns a `Decoder`.
+func NewDecoder(source []byte) *Decoder {
+	return &Decoder{
+		source: source,
+	}
+}
+
+func (d *Decoder) decode() ([]interface{}, error) {
+	return parse(d.source)
+}
+
+func parse(source []byte) (decoded []interface{}, err error) {
+	decoded = []interface{}{}
+
+	r := bytes.NewReader(source)
+
+	dec := yaml.NewDecoder(r)
+
+	dec.KnownFields(true)
+
+	// Iterate through all defined documents.
+	for {
+		var manifests yaml.Node
+
+		if err = dec.Decode(&manifests); err != nil {
+			if errors.Is(err, io.EOF) {
+				return decoded, nil
+			}
+
+			return nil, fmt.Errorf("decode error: %w", err)
+		}
+
+		if manifests.Kind != yaml.DocumentNode {
+			return nil, fmt.Errorf("expected a document")
+		}
+
+		for _, manifest := range manifests.Content {
+			var target interface{}
+
+			if target, err = decode(manifest); err != nil {
+				return nil, err
+			}
+
+			decoded = append(decoded, target)
+		}
+	}
+}
+
+//nolint: gocyclo
+func decode(manifest *yaml.Node) (target interface{}, err error) {
+	var (
+		version string
+		kind    string
+		spec    *yaml.Node
+	)
+
+	for i, node := range manifest.Content {
+		switch node.Value {
+		case ManifestVersionKey:
+			if err = manifest.Content[i+1].Decode(&version); err != nil {
```

Comment:
should we check that `i+1` is available? not sure, but what if document looks like `spec:` without any content on the right?

Reply:
Probably a good test case. I will add it.